utils/dataset.py

```diff
@@ -282,96 +412,65 @@ def get_binned_taco_dataloader(paths: Paths, max_batch_size: int = 8) -> BinnedT
     return BinnedTacoDataLoader(paths=paths, dataset=dataset, max_batch_size=max_batch_size)
 
 
-class TacoCollator:
+def _get_filtered_datasets(paths: Paths,
+                           max_mel_len: int,
+                           filter_duration_stats: bool,
+                           min_attention_alignment: float,
+                           min_attention_sharpness: float,
+                           max_consecutive_ones: int,
+                           max_duration: int) -> Tuple[List[tuple], List[tuple]]:
 
-    def __init__(self, r: int) -> None:
-        self.r = r
-
-    def __call__(self, batch: List[Dict[str, Union[str, torch.tensor]]]) -> Dict[str, torch.tensor]:
-        x_len = [b['x_len'] for b in batch]
-        x_len = torch.tensor(x_len)
-        max_x_len = max(x_len)
-        text = [pad1d(b['x'], max_x_len) for b in batch]
-        text = stack_to_tensor(text).long()
-        spec_lens = [b['mel_len'] for b in batch]
-        max_spec_len = max(spec_lens) + 1
-        if max_spec_len % self.r != 0:
-            max_spec_len += self.r - max_spec_len % self.r
-        mel = [pad2d(b['mel'], max_spec_len) for b in batch]
-        mel = stack_to_tensor(mel)
-        item_id = [b['item_id'] for b in batch]
-        speaker_name = [b['speaker_name'] for b in batch]
-        mel_lens = [b['mel_len'] for b in batch]
-        mel_lens = torch.tensor(mel_lens)
-        speaker_emb = [b['speaker_emb'] for b in batch]
-        speaker_emb = stack_to_tensor(speaker_emb)
+    train_data = unpickle_binary(paths.train_dataset)
+    val_data = unpickle_binary(paths.val_dataset)
+    speaker_dict = unpickle_binary(paths.speaker_dict)
 
-        return {'x': text, 'mel': mel, 'item_id': item_id,
-                'x_len': x_len, 'mel_len': mel_lens,
-                'speaker_emb': speaker_emb, 'speaker_name': speaker_name}
+    train_data = _filter_max_len(train_data, max_mel_len)
+    val_data = _filter_max_len(val_data, max_mel_len)
 
+    if filter_duration_stats:
+        duration_stats = unpickle_binary(paths.duration_stats)
+        data_filter = DataFilter(duration_stats=duration_stats,
+                                 min_attention_alignment=min_attention_alignment,
+                                 min_attention_sharpness=min_attention_sharpness,
+                                 max_consecutive_duration_ones=max_consecutive_ones,
+                                 max_duration=max_duration)
 
-class ForwardCollator:
+        speaker_counts_orig = Counter([speaker_dict[item_id] for item_id, _ in train_data + val_data if item_id in speaker_dict])
+        train_data = data_filter(train_data)
```

Comment:
While filtering train and val subsets there is a possible situation when distribution will be changed, for example, for certain speakers samples will be completely gone from train and kept in val or smth like that. But then the whole flow should be reviewed

Reply:
Could happen. I'd suggest to tackle this in a separate PR if needed.